redskyctl/internal/commands/run/form/form.go

```diff
@@ -0,0 +1,184 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package form
+
+import (
+	"strings"
+
+	"github.com/charmbracelet/bubbles/spinner"
+	"github.com/charmbracelet/bubbles/textinput"
+	tea "github.com/charmbracelet/bubbletea"
+)
+
+// Start initiates the start of a form. Note that there is no implicit scoping
+// of start messages to specific forms, proper routing of the start message is
+// the responsibility of the enclosing model.
+func Start() tea.Msg {
+	return startMsg{}
+}
+
+type startMsg struct{}
+
+// ValidationMsg is used to asynchronously validate form fields. An empty string
+// is used to indicate a valid field while any non-empty strings indicates an
+// error with the current field value. Validation messages are typically rendered
+// in the view of a field and will be cleared on the first key press after the
+// validation occurs.
+type ValidationMsg string
+
+// FinishedMsg indicates that the last field on the form has been submitted.
+type FinishedMsg struct{}
+
+// Fields is a list of fields that make up a form. When constructing a form, be
+// sure to use pointers to the individual field structs at the time of the update.
+// Most Bubble Tea updates operate on values, not references, so this is a little
+// different.
+type Fields []Field
+
+// Init returns a command to invoke in response to a form started message.
+func (f Fields) Init() tea.Cmd {
+	// In theory we could check to see if these are needed but it's easier to
+	// to fire off the single message and have it ignored.
+	return tea.Batch(textinput.Blink, spinner.Tick)
+}
+
+// Update manages the transition when individual fields are submitted.
+// IMPORTANT: the message cannot be delivered to the individual fields, they must
+// still be updated (presumably after the form itself has had a chance to update).
+func (f Fields) Update(msg tea.Msg) tea.Cmd {
+	focused, next := f.activeFields()
+
+	var cmds []tea.Cmd
+	switch msg := msg.(type) {
+
+	case startMsg:
+		// Focus the next field if nothing is currently focused
+		cmds = append(cmds, f.Init())
+		if focused == nil && next != nil {
+			next.Focus()
+			next.Show()
+		} else if focused != nil && focused.Hidden() {
+			focused.Show()
+		}
+
+	case tea.KeyMsg:
+		if focused != nil {
+			switch msg.Type {
+			case tea.KeyEnter:
+				// Validate the focused field when the user hits "enter"
+				// NOTE: A field like the textfield can also use `Validate` to effectively
+				// "block" a submission if there are pending actions that require "enter".
+				cmds = append(cmds, focused.Validate())
+
+			case tea.KeyPgUp:
```

Comment:
backspace key might be worth adding here as well.

Reply:
Backspace is tough because it might be handled by a text input as an actual backspace; we would need to special case based on the field (i.e. it would only really work for non-editable choice/multichoice fields).